### PR TITLE
Clipboard

### PR DESCRIPTION
**Pull request by munkk**

### CODE REVIEW

File: core/clipboard/clipboard.model.d.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import { Resource } from '../resource/resource';
```
Comment (klumba12) [inline core/clipboard/clipboard.model.d.ts]: remove import

File: core/clipboard/clipboard.model.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import {Resource} from '../resource';
+
+export class ClipboardModel {
+	constructor() {
```
Comment (klumba12) [inline core/clipboard/clipboard.model.js]: remove import and shortcut

File: core/clipboard/clipboard.view.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+export class ClipboardView extends View {
+	constructor(model, commandManager) {
+		super(model);
+
+		const selectionCommandManager = new SelectionCommandManager(model, commandManager);
+		const shortcut = model.action().shortcut;
+		const commands = this.commands;
+
+		const view = document.querySelector('.q-grid-view');
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: this is depricated in core, make a directive for that, or do it in view-core component

File: core/clipboard/clipboard.view.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+		view.setAttribute('onselectstart', 'return false');
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: can u add a comment in the code, what does it mean?

File: core/clipboard/clipboard.view.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+		return new Map(
+			Object.entries(commands)
+		);
+	}
+
+	handleColumn(items) {
+		const dataModel = this.model.data();
+		const rows = dataModel.rows;
+		const accumulator = [];
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: make a better name

File: core/clipboard/clipboard.view.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+				shortcut: shortcut.copy
+			})
+		};
+
+		return new Map(
+			Object.entries(commands)
+		);
+	}
+
+	handleColumn(items) {
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: make a better name, `handle` doesn't make sense

File: core/clipboard/clipboard.view.js [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+			})
+
+		});
+
+		return accumulator;
+	}
+
+	handleCell(items) {
+		const accumulator = [];
+		let collection = [];
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: better name

File: core/clipboard/clipboard.view.js [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+			});
+
+			accumulator.push(collection);
+		});
+
+		return accumulator;
+	}
+
+	buildTable(data) {
+		const table = document.createElement('table');
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: make a separate class for build and select table

File: core/clipboard/clipboard.model.d.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export declare class ClipboardModel {
+	constructor();
```
Comment (klumba12) [inline core/clipboard/clipboard.model.d.ts]: add shortcut here if it presents in clipboard model

Comment (klumba12) [reply core/clipboard/clipboard.model.d.ts]: Also please make tab separators for all files

File: core/clipboard/clipboard.service.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
+export class ClipboardService {
+
+	static buildTable(data) {
+		const table = document.createElement('table');
+		table.className = 'generatedTable';
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: `classList.add('q-grid...')`

File: core/clipboard/clipboard.service.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+		let body = document.body, range, sel;
+
+		if (document.createRange && window.getSelection) {
+			range = document.createRange();
+			sel = window.getSelection();
+			sel.removeAllRanges();
+			try {
+				range.selectNodeContents(el);
+				sel.addRange(range);
+			} catch (e) {
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: what does it code mean?

File: core/clipboard/clipboard.view.js [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+				shortcut: shortcut.copy
+			})
+		};
+
+		return new Map(
+			Object.entries(commands)
+		);
+	}
+
+	makeArrayFromColumns(items) {
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: what does mean makeArray? :)

File: core/clipboard/clipboard.view.js [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+				accumulator[celIndex][colIndex] = cells[celIndex];
+			})
+
+		});
+
+		return accumulator;
+	}
+
+	makeArrayFromCells(items) {
+		const accumulator = [];
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: bad name

File: core/clipboard/clipboard.view.js [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+		let collection = [];
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: bad name

File: core/clipboard/clipboard.service.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
+export class ClipboardService {
+
+	static buildTable(data) {
+		const table = document.createElement('table');
+		table.classList.add('q-grid-generated-table');
+
+		data.forEach((el) => {
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: what is data? what is o?

File: core/clipboard/clipboard.view.d.ts [hunk 16]
```diff
@@ -0,0 +1,8 @@
+import { View } from '../view/view';
+import { CommandManager } from '../command/command.manager';
+import { Model } from '../infrastructure/model';
+
+export declare class ClipboardView extends View {
+	constructor(model: Model, commandManager: CommandManager);
+}
```
Comment (klumba12) [inline core/clipboard/clipboard.view.d.ts]: Add copy command

File: core/clipboard/clipboard.view.js [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+			copy: new Command({
+				source: 'clipboard.view',
+				canExecute: () => selection.items.length > 0,
+				execute: () => {
+					const unit = selection.unit;
+					let items = selection.items;
+
+					switch (unit) {
+						case 'cell': {
+							items = this.buildFromCells(items);
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: rename all build* functions, and make more transparent names inside them

File: core/clipboard/clipboard.service.js [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+	document.body.appendChild(table);
+
+	return table;
+}
+
+function selectTable(element) {
+	let range, selection;
+
+	range = document.createRange();
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: const range = doc...
const selection = window....

File: core/clipboard/clipboard.service.js [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+		selectTable(table);
+	}
+}
+
+function buildTable(columns) {
+	const table = document.createElement('table');
+	table.classList.add('q-grid-clipboard');
+
+	columns.forEach((column) => {
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: please use `for(let i, i <  , i++` .. for perfomance

File: core/clipboard/clipboard.service.js [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+		const tr = document.createElement('tr');
+		for (let index in column) {
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: the same for loop with iterator

File: core/row/row.selector.js [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+
+	mapFromRows(rows) {
+		const result = [];
+		const columns = this.model.view().columns;
+
+		for(const row of rows) {
+			const line = [];
+
+			for (const column of columns) {
+				const label = getFactory(column);
```
Comment (klumba12) [inline core/row/row.selector.js]: add cache for this: `cache = new Map(); if (cache.has(column))...`

File: core/row/row.selector.js [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+				const value = label(row);
+				if(value && typeof value === 'string') {
+					line.push(value);
```
Comment (klumba12) [inline core/row/row.selector.js]: make it like `line.push(value === null || isUndefined(value) ? '' : '' + value)`

File: core/row/row.selector.js [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+	mapFromColumns(columns) {
+		const result = [];
+		const rows = this.model.view().rows;
+
+		columns.forEach((column, columnIndex) => {
+			const label = getFactory(column);
+			const cells = rows.map(row => label(row));
+
+			if (!result.length) {
+				cells.forEach((cell) => result.push(['empty']));
```
Comment (klumba12) [inline core/row/row.selector.js]: get rid of parenthesis `(cell => ...)

Comment (klumba12) [reply core/row/row.selector.js]: why we need `['empty']`?

File: core/row/row.selector.js [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+		});
+
+		return result;
+	}
+
+	mapFromCells(items) {
+		const result = [];
+		let line = [];
+
+		let keysForComparison = [];
```
Comment (klumba12) [inline core/row/row.selector.js]: what is the purpose for that?

File: core/row/row.selector.js [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+		result.push(line);
+
+		return result;
+	}
+
+	mapFromMix(items) {
+		for(const item of items) {
+
+			switch(item.unit) {
+				case 'row':
```
Comment (klumba12) [inline core/row/row.selector.js]: add braces in case stms: `case 'row': {...`

File: core/row/row.selector.js [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+					const row = item.item;
+					return this.mapFromRows([row]);
+				case 'cell':
+					const cells = [];
+					items.forEach(item => {
+						const itemRow = item.item.row;
```
Comment (klumba12) [inline core/row/row.selector.js]: rename it row and column and make `cells.push({row, column})`

File: core/utility/utility.js [hunk 27]
```diff
@@ -126,6 +126,10 @@ function escapeRegexp(text) {
 	return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
 }
 
+function getType(type) {
```
Comment (klumba12) [inline core/utility/utility.js]: remove this function

File: core/row/row.selector.js (hunk 23 above)
Comment (munkk) [reply core/row/row.selector.js]: I'm using here a two-dimensional array, and it first dimension should be initialized before starting to address the second. am i wrong?

File: core/row/row.selector.js (hunk 22 above)
Comment (munkk) [reply core/row/row.selector.js]: we will get extra columns if follow this method

File: core/row/row.selector.js (hunk 24 above)
Comment (munkk) [reply core/row/row.selector.js]: it helps accumulate keys of columns for further rows(lines) separation

File: core/row/row.selector.js (hunk 23 above)
Comment (klumba12) [reply core/row/row.selector.js]: put why not just `result.push([])`?

File: core/row/row.selector.js (hunk 24 above)
Comment (klumba12) [reply core/row/row.selector.js]: it's more convenient to group by cells by rows order by row indices and fill by columns

File: core/clipboard/clipboard.service.js [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+	const range = document.createRange();
+	const selection = window.getSelection();
+
+	selection.removeAllRanges();
+	range.selectNodeContents(element);
+	selection.addRange(range);
+
+	document.execCommand('copy');
+
+	const table = document.querySelector('.q-grid-clipboard');
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: remove should be outside of select

File: core/clipboard/clipboard.view.d.ts (hunk 16 above)
Comment (klumba12) [reply core/clipboard/clipboard.view.d.ts]: Copy command is not added still

File: core/clipboard/clipboard.view.js [hunk 29]
```diff
@@ -0,0 +1,49 @@
...
+		super(model);
+
+		const selectionCommandManager = new SelectionCommandManager(model, commandManager);
+
+		const shortcut = model.action().shortcut;
+		const commands = this.commands;
+
+		this.using(shortcut.register(selectionCommandManager, commands));
+	}
+
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: remove redundant whitespaces

File: core/row/row.selector.js [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+	mapFromRows(rows) {
+		const result = [];
+		const columns = this.model.view().columns;
+		const cache = new Map(columns.map(column => [column.key]));
+
+		for(const row of rows) {
+			const line = [];
+
+			for (const column of columns) {
+				if(cache.has(column.key)) {
```
Comment (klumba12) [inline core/row/row.selector.js]: dont get this code?

File: core/row/row.selector.js [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+					const label = getFactory(column);
+					const value = label(row);
+
+					if(value && typeof value === 'string') {
```
Comment (klumba12) [inline core/row/row.selector.js]: still don't get why we dont apply value if it is empty?

File: core/row/row.selector.js [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+		}
+
+		return result;
+	}
+
+	mapFromColumns(columns) {
+		const result = [];
+		const rows = this.model.view().rows;
+
+		columns.forEach((column, columnIndex) => {
```
Comment (klumba12) [inline core/row/row.selector.js]: for perfomance don't use .forEach when there can be lots of rows

File: core/row/row.selector.js [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+			const label = getFactory(column);
+			const cells = rows.map(row => label(row));
+
+			if (!result.length) {
+				cells.forEach(cell => result.push([]));
+			}
+
+			cells.forEach((cell, cellIndex) => {
```
Comment (klumba12) [inline core/row/row.selector.js]: use for(var i ...

File: core/row/row.selector.js [hunk 34]
```diff
@@ -0,0 +1,124 @@
...
+					const row = item.item;
+					return this.mapFromRows([row]);
+				}
+				case 'cell': {
+					const cells = [];
+					items.forEach(item => {
+						const row = item.item.row;
+						const column = item.item.column;
+
+						cells.push({row: row, column: column})
```
Comment (klumba12) [inline core/row/row.selector.js]: cells.push({row, column}) `row: row` is redundant

File: core/clipboard/clipboard.service.js [hunk 35]
```diff
@@ -0,0 +1,46 @@
+export class ClipboardService {
+	static copy(rows) {
+		const table = buildTable(rows);
+
+		selectTable(table);
+		removeTable();
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: Refactor to this
```javascript
const table = buildTable(rows);
select(table);
table.remove();
```

File: core/clipboard/clipboard.view.d.ts (hunk 16 above)
Comment (klumba12) [reply core/clipboard/clipboard.view.d.ts]: Still don't see copy command

File: core/clipboard/clipboard.view.js [hunk 36]
```diff
@@ -0,0 +1,46 @@
...
+	get commands() {
+		const model = this.model;
+		const selectionService = new SelectionService(model);
+		const rowSelector = new RowSelector(model);
+		const selectionState = model.selection();
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: you should invoke this everytime when command is executed

File: core/row/row.selector.js [hunk 37]
```diff
@@ -0,0 +1,134 @@
...
+				return this.mapFromCells(items);
+			case 'mix':
+				return this.mapFromMix(items);
+			default:
+				throw new AppError('row.selector', `Invalid unit ${selectionState.unit}`);
+		}
+	}
+
+	mapFromRows(rows) {
+		let value;
```
Comment (klumba12) [inline core/row/row.selector.js]: don't do like this, always try to make a variable scope small as possible

File: core/row/row.selector.js [hunk 38]
```diff
@@ -0,0 +1,134 @@
...
+			for (const column of columns) {
+				let label;
+				if(cache.has(column)) {
+					label = cache.get(column)
+				} else {
+					label = getFactory(column);
+					cache.set(column, label);
+				}
+
+				value = label(row);
```
Comment (klumba12) [inline core/row/row.selector.js]: `const value = ...`

File: core/row/row.selector.js [hunk 39]
```diff
@@ -0,0 +1,134 @@
...
+			.view()
+			.columns
+			.filter(column => column.class === 'data');
+
+		for (const row of rows) {
+			const line = [];
+
+			for (const column of columns) {
+				let label;
+				if(cache.has(column)) {
```
Comment (klumba12) [inline core/row/row.selector.js]: use `column.key`, it should work better and easier for debugging if needed

File: core/row/row.selector.js [hunk 40]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	mapFromRows(rows) {
+		let value;
+		const result = [];
+		const cache = new Map();
+		const columns = this.model
+			.view()
+			.columns
+			.filter(column => column.class === 'data');
```
Comment (klumba12) [inline core/row/row.selector.js]: add class 'pivot' to 'data' as an option

File: core/row/row.selector.js [hunk 41]
```diff
@@ -0,0 +1,134 @@
...
+		}
+
+		return result;
+	}
+
+	mapFromColumns(columns) {
+		const result = [];
+		const rows = this.model.view().rows;
+
+		for(let i = 0, max = columns.length; i < max; i++) {
```
Comment (klumba12) [inline core/row/row.selector.js]: rename max to columnLength

File: core/row/row.selector.js [hunk 42]
```diff
@@ -0,0 +1,134 @@
...
+			const column = columns[i];
+			const label = getFactory(column);
+			const cells = rows.map(row => label(row));
+
+			if (!result.length) {
+				for(let j = 0, max = cells.length; j < max; j++) {
```
Comment (klumba12) [inline core/row/row.selector.js]: rename max to cellsLength, this convension we try to use everywhere in the grid

File: core/row/row.selector.js [hunk 43]
```diff
@@ -0,0 +1,134 @@
...
+		return result;
+	}
+
+	mapFromColumns(columns) {
+		const result = [];
+		const rows = this.model.view().rows;
+
+		for(let i = 0, max = columns.length; i < max; i++) {
+			const column = columns[i];
+			const label = getFactory(column);
```
Comment (klumba12) [inline core/row/row.selector.js]: the same apply cache for this

File: core/row/row.selector.js [hunk 44]
```diff
@@ -0,0 +1,134 @@
...
+			for(let k = 0, max = cells.length; k < max; k++) {
+				result[k][i] = cells[k]
+			}
+		}
+
+		return result;
+	}
+
+	mapFromCells(items) {
+		const result = [];
```
Comment (klumba12) [inline core/row/row.selector.js]: will work on this in last attempt

File: core/row/row.selector.js [hunk 45]
```diff
@@ -0,0 +1,145 @@
...
+
+		return result;
+	}
+
+	mapFromColumns(columns) {
+		const result = [];
+		const rows = this.model.view().rows;
+		const cache = new Map();
+
+		for(let i = 0, columnLength = columns.length; i < columnLength; i++) {
```
Comment (klumba12) [inline core/row/row.selector.js]: make a method `mapFromRowColumns(rows, columns)`
and use there `mapFromRows` body
inside `mapFromRows` and `mapFromColumns` use mapFromRowColumns

File: core/row/row.selector.d.ts [hunk 46]
```diff
@@ -0,0 +1,16 @@
...
+
+export declare class RowSelector {
+	constructor(model: Model);
+
+	model: Model;
+	source: string[];
+
+	map(items: any[]): void;
+	mapFromRows(rows: any[]): (rows: any[], columns: any[]) => any;
+	mapFromColumns(columns: any[]): (rows: any[], columns: any[]) => any;
```
Comment (titov-max) [inline core/row/row.selector.d.ts]: Looks that ```columns``` is array of ```ColumnModel``` in these methods

Comment (munkk) [reply core/row/row.selector.d.ts]: right)

File: core/clipboard/clipboard.service.js [hunk 47]
```diff
@@ -0,0 +1,77 @@
...
+		const table = buildTable(rows, head, foot);
+		select(table);
+		table.remove();
+	}
+}
+
+function buildTable(rows, head, foot) {
+	const table = document.createElement('table');
+	table.classList.add('q-grid-clipboard');
+	let headFlag = true;
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: What are the flags? what they indicate?

File: core/clipboard/clipboard.service.js [hunk 48]
```diff
@@ -0,0 +1,77 @@
...
+
+export class ClipboardService {
+	static copy(rows, head, foot) {
+		const table = buildTable(rows, head, foot);
+		select(table);
+		table.remove();
+	}
+}
+
+function buildTable(rows, head, foot) {
```
Comment (klumba12) [inline core/clipboard/clipboard.service.js]: `buildTable(selectors) {}`

`javascript
selector = {
rows: rows from head/foot/body,
tag: th, td
}
`

File: core/clipboard/clipboard.view.js [hunk 49]
```diff
@@ -0,0 +1,79 @@
...
+					const selectionService = new SelectionService(this.model);
+					const rowSelector = new RowSelector(this.model);
+					const selectionItems = selectionState.items;
+					const source = this.model.clipboard().source.join(', ').toLowerCase();
+					const entries = selectionService.lookup(selectionItems);
+					const rows = rowSelector.map(entries);
+					const head = rows.shift();
+					const foot = rows.pop();
+
+					switch (source) {
```
Comment (klumba12) [inline core/clipboard/clipboard.view.js]: adjust to new buildTable interface

File: core/row/row.selector.js [hunk 50]
```diff
@@ -0,0 +1,158 @@
+import {AppError} from '../infrastructure';
+import {getFactory, get} from '../services/label';
+import {isUndefined} from '../utility/utility';
+import {Aggregation} from '../services';
+import {getFactory as valueFactory} from '../services/value';
+
+export class RowSelector {
+	constructor(model) {
```
Comment (klumba12) [inline core/row/row.selector.js]: head and foot shouldn't be in this file

**Pull request closed without merging**